atomic_reactor/util.py

```diff
@@ -1672,3 +1672,177 @@ def chain_get(d, path, default=None):
+            return SCRATCH_FROM
+
+        if self._pullable_parents and self._pullable_parents[0]:
+            return deepcopy(self._pullable_parents[0])
+        else:
+            raise KeyError('base_image')
+
+    @property
+    def base_from_scratch(self):
+        return self._base_from_scratch
+
+    @property
+    def custom_base_image(self):
+        return self._custom_base_image
+
+    @property
+    def custom_parent_image(self):
+        return self._custom_parent_image
+
+    @property
+    def source_registry(self):
+        return self._source_registry
+
+    @property
+    def organization(self):
+        return self._organization
+
+    def __len__(self):
+        return len(self._pullable_parents)
+
+    def __setitem__(self, key, item):
+        if not isinstance(key, ImageName):
+            key = ImageName.parse(key)
+
+        if not base_image_is_custom(key.to_str()):
+            if not key.registry:
+                key.registry = self._source_registry
+
+            if self._should_enclose(key):
+                key.enclose(self._organization)
+
+        try:
+            kindex = self._pullable_parents.index(key)
+        except ValueError:
+            raise KeyError(key)
+
+        self._local_parents[kindex] = ImageName.parse(item)
+        logger.info("set parent image '%s' to '%s'", key.to_str(), item)
+
+    def __getitem__(self, key):
+        if not isinstance(key, ImageName):
+            key = ImageName.parse(key)
+
+        if not base_image_is_custom(key.to_str()):
+            if not key.registry:
+                key.registry = self._source_registry
+
+            if self._should_enclose(key):
+                key.enclose(self._organization)
+
+        try:
+            kindex = self._pullable_parents.index(key)
+        except ValueError:
+            raise KeyError(key)
+        return deepcopy(self._local_parents[kindex])
+
+    def __iter__(self):
+        return iter(self.keys())
+
+    def keys(self):
+        return deepcopy(self._pullable_parents)
+
+    def values(self):
+        return deepcopy(self._local_parents)
+
+    def items(self):
+        return zip(self.keys(), self.values())   # pylint: disable=zip-builtin-not-iterating
+
+    def _create_pullable(self):
+        for index, image in enumerate(reversed(self._original_parents)):
+            image_name = ImageName.parse(image)
+            if base_image_is_scratch(image_name.get_repo()):
+                if index == 0:
+                    self._base_from_scratch = True
+                continue
+            if base_image_is_custom(image_name.to_str()):
+                self._custom_parent_image = True
+                if index == 0:
+                    self._custom_base_image = True
+
+            if image_name not in self._pullable_parents:
+                self._pullable_parents.append(image_name)
+                self._local_parents.append(None)
+
+    def set_source_registry_and_org(self, source_registry, organization=None):
```

Comment:
Methods having `_and_` in name are suspicious. (Single responsibility principle)

You have already properties for source_registry and organization, so this should be split into two setters

Reply:
reason for having just 1 method is that we will set it just once, so 1 method seemed enough,
but if you insist I can split it

Reply:
Is there a good name for the combined source_reg+org? If so, then this becomes `set_FOO()`